File: app/interface/http/schemas/weather.py
"""Raw (normalized) weather schemas — API Spec §9.9 and the §8.4 payload.

Deliberately omits `completeness` and `source_class`: those are *internal*
normalization fields (they feed `travelConfidence`), and §9.9 does not
publish them. No provider is identified anywhere here.
"""

from datetime import date

from app.domain.entities.weather import NormalizedReading
from app.domain.entities.weather_intelligence import Period, ResolvedLocation
from app.interface.http.schemas.common import CamelModel, LocationSchema, PeriodSchema
from app.interface.http.schemas.intelligence import location_from_domain, period_from_domain


class RawWeatherReadingSchema(CamelModel):
    """API Spec §9.9."""

    date: date
    temp_min_c: float
    temp_max_c: float
    precipitation_probability: float
    precipitation_mm: float | None = None
    wind_speed_kph: float
    humidity: float | None = None
    condition: str

    @classmethod
    def from_domain(cls, reading: NormalizedReading) -> "RawWeatherReadingSchema":
        return cls(
            date=reading.date,
            temp_min_c=reading.temp_min_c,
            temp_max_c=reading.temp_max_c,
            precipitation_probability=reading.precipitation_probability,
            precipitation_mm=reading.precipitation_mm,
            wind_speed_kph=reading.wind_speed_kph,
            humidity=reading.humidity,
            condition=reading.condition.value,
        )


class RawWeatherViewSchema(CamelModel):
    """Payload of `GET .../weather/raw` (API Spec §8.4)."""

    location: LocationSchema
    period: PeriodSchema
    readings: list[RawWeatherReadingSchema]

    @classmethod
    def from_domain(
        cls, location: ResolvedLocation, period: Period, readings: list[NormalizedReading]
    ) -> "RawWeatherViewSchema":
        return cls(
            location=location_from_domain(location),
            period=period_from_domain(period),
            readings=[RawWeatherReadingSchema.from_domain(r) for r in readings],
        )

File: frontend/src/types/weather.ts
/** Weather and per-day intelligence — API Spec §9.4, §9.9. */

import type { IsoDate, Score, UnitInterval } from './common'
import type { Location, Period } from './location'
import type {
  ActivityCategory,
  RiskFactorType,
  RiskLevel,
  Severity,
  TravelAdvisory,
  WeatherCondition,
} from './enums'

/** API Spec §9.4.1. Units are metric and fixed in v1; °F is a client-side conversion. */
export interface DailySummary {
  tempMinC: number
  /** Always ≥ tempMinC. */
  tempMaxC: number
  /** 0.0–1.0. Rendered as a percentage. */
  precipitationProbability: UnitInterval
  windSpeedKph: number
  condition: WeatherCondition
}

/**
 * API Spec §9.4.3. The explainability anchor (NFR-1).
 *
 * `description` is the human-facing text and is always shown on expansion.
 * `rule` is the developer-facing id — available on demand, never primary
 * (FDS §8.5).
 */
export interface RiskFactor {
  type: RiskFactorType
  severity: Severity
  description: string
  rule: string
}

/** API Spec §9.4.2. `riskFactors` may legitimately be empty on a low-risk day. */
export interface RiskAssessment {
  overallRiskLevel: RiskLevel
  riskFactors: RiskFactor[]
}

/** API Spec §9.4.4. */
export interface ActivitySuitability {
  activity: ActivityCategory
  /** Integer 0–100. Higher is more suitable. */
  score: Score
}

/** API Spec §9.4. One entry per day in the requested range, ascending by date. */
export interface DailyIntelligence {
  date: IsoDate
  summary: DailySummary
  riskAssessment: RiskAssessment
  activitySuitability: ActivitySuitability[]
  packingRecommendations: string[]
  travelAdvisory: TravelAdvisory
}

/**
 * API Spec §9.9. Underlying normalized reading, without intelligence.
 *
 * `precipitationMm` and `humidity` are genuinely nullable — a provider may
 * simply not report them. They render as "—", never as `0` (FDS §6.5).
 */
export interface RawWeatherReading {
  date: IsoDate
  tempMinC: number
  tempMaxC: number
  precipitationProbability: UnitInterval
  precipitationMm?: number | null
  windSpeedKph: number
  /** 0.0–1.0 relative humidity, if available. */
  humidity?: UnitInterval | null
  condition: WeatherCondition
}

/** Payload of `GET /locations/{id}/weather/raw` — API Spec §8.4. */
export interface RawWeatherView {
  location: Location
  period: Period
  readings: RawWeatherReading[]
}
